dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorPower.java

```diff
@@ -285,4 +299,336 @@ public List<Event> visitDat3mCAS(Dat3mCAS e) {
+		String mo = e.getMo();
+
+		ExprInterface expected = e.getCmp();
+        Label casEnd = newLabel("CAS_end");
+        CondJump branchOnCasCmpResult = newJump(new Atom(resultRegister, NEQ, expected), casEnd);
+
+        // Power does not have mo tags, thus we use null
+        Load loadValue = newRMWLoadExclusive(resultRegister, address, null);
+        Store storeValue = newRMWStoreExclusive(address, value, null, true);
+        Label label = newLabel("FakeDep");
+        Event fakeCtrlDep = newFakeCtrlDep(resultRegister, label);
+
+        Fence optionalMemoryBarrierBefore = mo.equals(Tag.Linux.MO_MB) ? Power.newSyncBarrier()
+                : mo.equals(Tag.Linux.MO_RELEASE) ? Power.newLwSyncBarrier() : null;
+        Fence optionalMemoryBarrierAfter = mo.equals(Tag.Linux.MO_MB) ? Power.newSyncBarrier()
+        		: mo.equals(Tag.Linux.MO_ACQUIRE)? Power.newISyncBarrier() : null;
+
+        return eventSequence(
+                // Indentation shows the branching structure
+                optionalMemoryBarrierBefore,
+                loadValue,
+                branchOnCasCmpResult,
+                    storeValue,
+                    fakeCtrlDep,
+                    label,
+                    optionalMemoryBarrierAfter,
+                casEnd
+        );
+	}
+	
+	@Override
+	public List<Event> visitRMWXchg(RMWXchg e) {
+		Register resultRegister = e.getResultRegister();
+		ExprInterface value = e.getMemValue();
+		IExpr address = e.getAddress();
+		String mo = e.getMo();
+
+        // Power does not have mo tags, thus we use null
+        Load load = newRMWLoadExclusive(resultRegister, address, null);
+        Store store = newRMWStoreExclusive(address, value, null, true);
+        Label label = newLabel("FakeDep");
+        Event fakeCtrlDep = newFakeCtrlDep(resultRegister, label);
+
+        Fence optionalMemoryBarrierBefore = mo.equals(Tag.Linux.MO_MB) ? Power.newSyncBarrier()
+                : mo.equals(Tag.Linux.MO_RELEASE) ? Power.newLwSyncBarrier() : null;
+        Fence optionalMemoryBarrierAfter = mo.equals(Tag.Linux.MO_MB) ? Power.newSyncBarrier()
+        		: mo.equals(Tag.Linux.MO_ACQUIRE)? Power.newISyncBarrier() : null;
+
+        return eventSequence(
+                optionalMemoryBarrierBefore,
+                load,
+                store,
+                fakeCtrlDep,
+                label,
+                optionalMemoryBarrierAfter
+        );
+	}
+	
+	@Override
+	public List<Event> visitRMWOp(RMWOp e) {
+		Register resultRegister = e.getResultRegister();
+		IOpBin op = e.getOp();
+		IExpr value = (IExpr) e.getMemValue();
+		IExpr address = e.getAddress();
+		String mo = e.getMo();
+		
+        Register dummyReg = e.getThread().newRegister(resultRegister.getPrecision());
+        Local localOp = newLocal(dummyReg, new IExprBin(resultRegister, op, value));
+
+        // Power does not have mo tags, thus we use null
+        Load load = newRMWLoadExclusive(resultRegister, address, null);
```

Comment:
`RMWOp` is the non-returning counterpart of `RMWOpReturn`, right? So technically, it should not have a `resultRegister` to begin with since it returns nothing (I guess `RMWOp` has a result register because it inherits it?).
I don't mind using the `resultRegister` as a dummy here, since no-one is gonna read it, but we should have at least a comment saying that this is supposed to not return anything and that `resultRegister` is just used as a dummy.